138 subproject template

dasdranagon opened the pull request:
This is a template project which can be copied to start working on a new component.
I've considered creating a template project for IDE, but it is not working as well as I expected and is not sharing between Android Studio and AppCode. So this is a simple and working solution. I'll update wiki with the "Contribute to Kaluga" workflow as soon as these changes will be merged.

avdyushin commented on template/README.md:
```diff
@@ -0,0 +1,35 @@
+This project includes all common dependencies and settings for kaluga subproject.
+
+#Steps for adding new subproject
+1. CopyPaste `teplete` subproject, select name for your subproject (we recommend to use only lowercase characters).
```
typo `teplete`

avdyushin commented on template/README.md:
```diff
@@ -0,0 +1,35 @@
+
+#Steps for adding new subproject
+1. CopyPaste `teplete` subproject, select name for your subproject (we recommend to use only lowercase characters).
+2. `<your subproject/>src/androidLibMain/AndroidManifest.xml` change `com.splendo.kaluga.template` to your subproject package name.	
```
I guess should be `<your subproject>/src/...`

avdyushin commented on template/README.md:
```diff
@@ -0,0 +1,35 @@
+	`include(":<your subproject>")`.
+6. Add your source code. For more derails check README.md in source root and each platform folder.
+7. Remove this markdown file from your subproject.
+8. Don't forget to add copyright if files were imported from another project
```
FYI, this copyright template already embedded into project and you can easily update 'licence' for selected files

avdyushin commented on template/src/commonTest/kotlin/README.md:
```diff
@@ -0,0 +1,12 @@
+5. In a field "Tasks" enter cleanAllTests allTests.
+6. Press "Ok" or "Apply". Now you can choose and run configuration.
+
+Please note that this scheme will run all gradle test tasks including iOS. Use [test filtering](https://docs.gradle.org/current/userguide/java_testing.html#test_filtering)   if you want tu run some specific tests.   
```
Extra spacing and typo in `tu`

avdyushin commented on template/src/iosMain/kotlin/README.md:
```diff
@@ -0,0 +1,11 @@
+5. In a field "Tasks" enter cleanIosx64Test iosx64Test.
+6. Press "Ok" or "Apply". Now you can choose and run configuration.
+
+Use [test filtering](https://docs.gradle.org/current/userguide/java_testing.html#test_filtering)   if you want tu run some specific tests.   
```
Also to many spaces and `tu`

dasdranagon replied:
😂

dasdranagon commented:
Thanks, @avdyushin. Fixing...

avdyushin commented on template/README.md:
```diff
@@ -0,0 +1,17 @@
+This project includes all common dependencies and settings for kaluga subproject.
+
+#Steps for adding new subproject
+1. CopyPaste `teplate` subproject, select name for your subproject (we recommend to use only lowercase characters).
```
Typo still here, must be CopyPaste `template` subproject, ...

avdyushin commented on template/README.md:
```diff
@@ -0,0 +1,17 @@
+7. Remove this markdown file from your subproject.
+8. Don't forget to update copyright if files were imported from another project. To do that use **Code > Update Copyright...**.
+9. Add documentation to your code if needed.
+10. Add an example opf usage to the example project.
```
Also typo in `opf`

dasdranagon commented:
Fixed typo, checked using cSpell and pushed

thoutbeckers commented on template/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+    mavenCentral()
+}
+
+dependencies {
+    implementation("androidx.fragment:fragment:1.1.0")
+    implementation("androidx.appcompat:appcompat:1.1.0")
+    testImplementation("org.mockito:mockito-core:2.28.2")
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.3.1")
+    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.3.1")
+    androidTestImplementation("androidx.test.uiautomator:uiautomator:2.2.0")
+    androidTestImplementation("androidx.fragment:fragment-ktx:1.1.0")
+}
+
```
Some of these are already in `component.gradle`, remove dupes.

thoutbeckers replied:
Or rather, if some of these are so needed, they should be in components.gradle to begin with. But e.g. I don't agree every project needs fragments.

dasdranagon replied:
Moved it to component.gradle, disabled fragment. But, because fragments is quite common used dependency I still prefer to have it commented in templete.

thoutbeckers replied:
commented in is a good compromise (in the template then of course). Shall we add a commented line `You can uncomment these frequently used dependencies if you need one` or something` to explain it a bit.

thoutbeckers commented on template/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+
+android {
+    testOptions {
+        unitTests.isReturnDefaultValues = true
```
Already in `component.gradle`

thoutbeckers commented on template/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+kotlin {
+    targets {
+        tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
+            kotlinOptions {
+                jvmTarget = "1.8"
+            }
+        }
+    }
+
```
Already in `component.gradle`

thoutbeckers commented on template/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+
+                val ext = (gradle as ExtensionAware).extra
+                implementation(project(":base", ""))
+                implementation("co.touchlab:stately-common:${ext["stately_version"]}")
+                implementation("co.touchlab:stately-concurrency:${ext["stately_version"]}")
```
Should only be added if you use them.

Putting `base` here seems like a good idea though.

dasdranagon replied:
Disabled  (Commented) this part

thoutbeckers replied:
good idea

thoutbeckers commented on template/src/androidLibAndroidTest/kotlin/README.md:
```diff
@@ -0,0 +1,43 @@
+## Android instrumented unit tests
```
Do we want the same README in every component?

thoutbeckers replied:
This readme will be out of date really quickly.

dasdranagon replied:
Probably it is not a common case, but for me as for iOS developer took some it took some time to figure out how to start writing tests and why for android there are two different folders. I understand that you can read a few articles and take a look at the official documentation, but I would find it more convenient to have such quick HOWTO. 
In general, I just wrote down all questions I had trying to contribute to Kaluga.

thoutbeckers replied:
I'm not opposed to the READMEs, and if I want to improve them I shall make a PR myself. I'm against them being copied into every new module, where we cannot properly maintain them.

I would be for making the template folder level one deeper (e.g. `/adding-a-new-module/template`, or putting the even putting one or more markdown files  in the root of the whole project (since they are quite important)

thoutbeckers commented on template/README.md:
```diff
@@ -0,0 +1,17 @@
+7. Remove this markdown file from your subproject.
+8. Don't forget to update copyright if files were imported from another project. To do that use **Code > Update Copyright...**.
+9. Add documentation to your code if needed.
+10. Add an example of usage to the example project.
```
There is a step 11 (if you have instrumentation tests for Android), but I can make a PR it.

dasdranagon replied:
Not sure that I understand which step. For me, it works. So yes, please make PR or let me know what can I add (Didn't research good enough instrumented tests)

thoutbeckers commented on template/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+}
+
+dependencies {
+    implementation("androidx.fragment:fragment:1.1.0")
+    implementation("androidx.appcompat:appcompat:1.1.0")
+    testImplementation("org.mockito:mockito-core:2.28.2")
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.3.1")
+    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.3.1")
+    androidTestImplementation("androidx.test.uiautomator:uiautomator:2.2.0")
+    androidTestImplementation("androidx.fragment:fragment-ktx:1.1.0")
```
Version numbers are hardcoded too.

dasdranagon replied:
Everything was copypaste from HUD and Alerts. I'll check all submodules and make changes if needed

thoutbeckers replied:
you were looking at (quite) old versions of those, as you can see in master they don't contain many of the things you put in (because they are in `component.gradle` or obsolete).

dasdranagon commented:
@tijl and @avdyushin  I've updated Android dependencies, made some small update in readme and move the template project to adding-a-new-module folder

thoutbeckers commented on adding-a-new-module/template/README.md:
```diff
@@ -0,0 +1,17 @@
+## Template 
```
The goal of having a subdirectory is so the README guides can live there, and they're not copied each time into a new module. They're all still in the same place however, so will end up copied.

dasdranagon replied:
Yes. Added in notes that that should be removed. I would remove it from the project as well but I just afraid that git will not understand empty directory and that will not appear in a project if you clone it. I would prefer rather have some code with garbage witch compiles out of the box than search in a wiki what should I do to make my code working.  

It is quite obvious anyway to keep your code clean and clean code before publishing.

thoutbeckers replied:
This particular file is not in an empty directory, so it should move one up, then people don't have to delete it and see it as the welcome file when they click on `add-a-new-module`.

For the empty dir trick, see comments there, we can keep that in some form.

dasdranagon replied:
Okey. now I understand what you mean. I'll move it

thoutbeckers commented on adding-a-new-module/template/build.gradle.kts:
```diff
@@ -0,0 +1,47 @@
+group = "com.splendo.kaluga"
+version = ext["library_version"]!!
+
+repositories {
+    google()
+    mavenCentral()
+}
```
also in `components.gradle`, can probably be cleaned up from modules where it still is added.

dasdranagon replied:
Yep. But I would prefer to do it in separate PR (clean other projects)

thoutbeckers replied:
Good idea to do those separate, but can be removed from the template.

thoutbeckers commented on adding-a-new-module/template/build.gradle.kts:
```diff
@@ -0,0 +1,47 @@
+
+dependencies {
+    /* Uncomment this line if you are using mockito
+     testImplementation("org.mockito:mockito-core:2.28.2")
```
Hardcoded version (if something will be shared by multiple modules or with example project we want to use a var), but also not needed I assume? 

`android_dependencies.gradle` has `org.mockito:mockito-core:3.1.0'` (note that since it is the only place this is declared, no var is needed per se)

thoutbeckers commented on adding-a-new-module/template/build.gradle.kts:
```diff
@@ -0,0 +1,47 @@
+    */
+
+    /* Uncomment these lines if you are using fragments
+    implementation("androidx.fragment:fragment:1.1.0")
```
hardcoded version,

Since we will copy this everywhere can be simplified to `implementation androidx.fragment:fragment-ktx:$whatever`, that should have `androidx.fragment:fragment` as a dependency and are strongly promoting kotlin.

dasdranagon replied:
As I understand "androidx.fragment:fragment-ktx" used for test implementation only so I prefer to leave it there

thoutbeckers replied:
No it stands for *K*o*t*lin E*x*tensions. There are many of these, they add some nice kotlin methods to java based packages. 

Someone needed it in a test so added it in the test scope, then everyone copy pasted it.

dasdranagon replied:
Okey. I'll remove it

thoutbeckers replied:
Ok, for the record I did not suggest removing it.

dasdranagon replied:
Okey. It's commented anyway ...

thoutbeckers replied:
My suggestion (first comment) is to simplify it to a single line:

`androidx.fragment:fragment-ktx:$whatever`

dasdranagon replied:
Done

thoutbeckers commented on adding-a-new-module/template/src/androidLibAndroidTest/kotlin/TestActivity.kt:
```diff
@@ -0,0 +1,34 @@
+class TestActivity : AppCompatActivity() {
+    // val viewModel: <Concrete ViewModel here> by viewModels()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+    }
```
Not sure what the benefit of pre declaring these are (espc onDestroy)

dasdranagon replied:
I'll remove onDestroy. onCreate is probably needed in 99.99(9)% of cases.

thoutbeckers replied:
Thinking about this more, you often don't need an Activity at all (at least `location`, `logging`, `test-utils` don't, and two do have tests here).  Should this even be in here?

dasdranagon replied:
Yep, sometimes we don't need instrumented tests, but I would prefer to live it in a template. Delete is allways easier than add

thoutbeckers replied:
Template for instrumented tests is fine (to have the directories etc) but specifically having an Activity goes a bit far. It's just a plain Activity.

It's also missing from the manifest incidently.

thoutbeckers replied:
onRemove is now gone, but now there's just an empty activity (not in any manifest still right?)

dasdranagon replied:
should be in /androidLibAndroidTest/AndroidManifest.xml

thoutbeckers commented on adding-a-new-module/template/src/commonMain/kotlin/README.md:
```diff
@@ -0,0 +1 @@
+## Here  common sources
```
Can we not just put in the main README (for adding a new module) what these paths mean?

dasdranagon replied:
I would prefer to have something in this directory. would like to have a project structure created (don't only describe) in the template - it's why it calls template.
Not sure how the Gradle project will behave if it will have reference to a not existing directory

thoutbeckers replied:
Ah the empty directory problem.

If we have a guide in /adding-a-new-module/something.md we could have these file and link to (a section in) that file. That way the README has 2 uses, it points you to the right central documentation, and it keeps the directory.

Of course a single of text before the link (as you have in this file) is not a bad idea either, it need a little more work than `Here  common sources` though.

thoutbeckers replied:
Not sure why the README files are now removed after I agreed to keep them because of the empty directory problem?

avdyushin replied:
`touch .keep` ?

dasdranagon replied:
As I understand you wont to remove them or update text inside. So I'm not sure what can I add in a file that exists only to be deleted, so just removed it.
Good point from @avdyushin

thoutbeckers replied:
`.keep` is a common solution, but please read carefully what I actually commented on this. A small README with a link to a section of the main README is perfectly fine with me.

This is what we need to do for the larger READMEs anyway, and would solve both problems (keeping the directory, and explaining what it is for) without creating long copy/pasted text everywhere that goes out of date.

dasdranagon replied:
Okey. Yes, it sounds good. I'll update README

thoutbeckers commented on gradle/android_dependencies.gradle:
```diff
@@ -25,7 +25,11 @@ dependencies {
     testImplementation "org.jetbrains.kotlinx:kotlinx-coroutines-test:$gradle.kotlinx_coroutines_version"
     testImplementation 'org.jetbrains.kotlin:kotlin-test'
     testImplementation 'org.jetbrains.kotlin:kotlin-test-junit'
+    testImplementation "org.junit.jupiter:junit-jupiter-api:$gradle.android_test_junit_jupiter_version"
 
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$gradle.android_test_junit_jupiter_version"
+
```
Does something actually need this?

dasdranagon replied:
Yes. I don't have that much experience with Android so I just copypasted from another module, but I was not able to run android (not instrumented) unit tests without it. idea doesn't just sais "No tests are found"

thoutbeckers replied:
This is for your own tests? (as you can see on CI all the other modules have no problem running). 

Android tests are JUnit4 based by default, this seems JUnit5 related.

dasdranagon replied:
Just added some tests and idea doesn't see them. Don't know why, but okey, I'll remove it

thoutbeckers replied:
give me the branch and I can help figure out why.

dasdranagon replied:
There is no branch, I just add some dummy sources and tests to this project to check that something that we recommend to use as a template works. So you can clone this branch and add some simple class ( class Dummy(val description: String) android unit tests following instructions in README. In my case, IDE cannot find Android unit tests. Of course, it could be some problem on my local machine, so I just removed it form android_dependencies.

thoutbeckers replied:
If adding this fixes it for you, there's probably still something wrong with the way you wrote your test, or something is broken. 

I noticed android integration tests are not on by default on Bitrise yet, I'll enable them.

thoutbeckers commented on gradle/android_dependencies.gradle:
```diff
@@ -25,7 +25,11 @@ dependencies {
 
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$gradle.android_test_junit_jupiter_version"
+
+    androidTestImplementation 'androidx.test.uiautomator:uiautomator:2.2.0'
```
Already declared 5 lines below.

dasdranagon replied:
Okey I'll remove automator

dasdranagon commented:
@thoutbeckers and @avdyushin I've pushed some changes:
1. Add version for fragment dependencies
2. Removed mockito from template dependencies
3. Removed empty onDestroy from TestActivity.

I still think it is better to leave README files in each platform directory. I tried to remove them and clone the project in a clean directory and as I expected all empty folders didn't appear in the project. So prefer to leave them so that we will have prepared structure in the template.

thoutbeckers commented:
@dasdranagon you seemed to have missed some of my comment (e.g. on empty dirs). Please check you read them all.

thoutbeckers commented:
We should also choose, is it a "subproject" or is it a "module"?

thoutbeckers commented:
Integration tests now fail, seems due to the previous PR perhaps, not sure. We can fix that in master first.

dasdranagon commented:
@thoutbeckers and @avdyushin  I updated the main README file, removed .keep files, and added README instead with a link to a section the main documentation file

thoutbeckers commented on adding-a-new-module/README.md:
```diff
@@ -0,0 +1,128 @@
+## Template 
+This **project** includes all common dependencies and settings for kaluga module.
+
+#Steps for adding new module
```
missing space

thoutbeckers commented on adding-a-new-module/README.md:
```diff
@@ -0,0 +1,128 @@
+11. Add an example of usage to the example project.
+
+---
+### <a name="commonTests"></a> Common Unit Tests 
```
FYI anchors are generated by GitHub based on the title of a heading automatically

thoutbeckers replied:
(no problem to leave them in however, these are technically more stable)

thoutbeckers commented on adding-a-new-module/template/build.gradle.kts:
```diff
@@ -0,0 +1,37 @@
+    /* Uncomment these lines if you are using fragments
+    val ext = (gradle as ExtensionAware).extra
+    androidTestImplementation("androidx.fragment:fragment-ktx:${ext["androidx_fragment_version"]}")
+*/
```
missing whitespace

thoutbeckers commented on .idea/codeStyles/codeStyleConfig.xml:
```diff
@@ -1,5 +1,6 @@
 <component name="ProjectCodeStyleConfiguration">
   <state>
     <option name="USE_PER_PROJECT_SETTINGS" value="true" />
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default" />
```
Not blocking for me, IDEA always tries to add it, but @avdyushin does this affect your code style settings somehow? Or is it just ignored? Does Android Studio try to add it back in again?

dasdranagon replied:
Sorry, I didn't notice. Removed it

avdyushin replied:
Well, we should use code style from one attached to the project itself (I hope use per project stays for). But I have no idea why IDEA/AS keep updating settings back to default

thoutbeckers commented:
Merged since this is not what broke master anyway. Now you can propose your main README changes @dasdranagon .
